working/0546-patterns/patterns-feature-specification.md

```diff
@@ -541,24 +540,40 @@ is expected.
 ### List pattern
 
 ```
-listPattern ::= typeArguments? '[' patterns? ']'
+listPattern         ::= typeArguments? '[' listPatternElements? ']'
+listPatternElements ::= listPatternElement ( ',' listPatternElement )* ','?
+listPatternElement  ::= pattern | restPattern
+restPattern         ::= '...' pattern?
 ```
 
 A list pattern matches an object that implements `List` and extracts elements by
 position from it.
 
-It is a compile-time error if `typeArguments` is present and has more than one
-type argument.
+It is a compile-time error if:
+
+*   `typeArguments` is present and has more than one type argument.
+
+*   There is more than one `restPattern` element in the list pattern. *It can
+    appear anywhere in the list, but there can only be zero or one.*
 
-**TODO: Allow a `...` element in order to match suffixes or ignore extra
-elements. Allow capturing the rest in a variable.**
+#### Rest patterns
+
+A list pattern may contain a *rest element* which allows matching lists of
+arbitrary lengths. If a rest element is present and has a subpattern, all of the
+elements not matched by other subpatterns are collected into a new list and that
+list is matched against the rest subpattern.
```

Comment:
Do we really want to specify that a new list is allocated?  Should we consider a using "slice" of the underlying list instead?

Reply:
Personally I'm ok with `sublist()`.  It's simple and well-defined, and reasonably performant for the cases that I believe are going to arise in practice.

I might have been ok with a slice if slices were a thing that were commonly used in Dart, but they aren't.  And because of that, I think people are going to assume sublist-like semantics with regard to mutation (e.g. `var [x, ...rest] = y; rest.add(0);` doesn't modify the original list).

I don't want to go to extra implementation effort to optimize a list pattern appearing to the right of `...` because I don't think anyone is going to write code like that.  Our time is better spent on other language features.  If someone really does write `[var [var a, ...[var b, ...var r, var c], var d] = [1,2,3,4,5];` and they get sub-optimal performance, well, they asked for it.

Reply:
I agree with everything Paul said. It felt a little arbitrary when I first thought to use `sublist()`, but the longer I thought about it, the more I like it. It *is* part of the `List` interface, so it's right there for the taking. It makes it easier to explain to users what kind of object they get if the rest pattern captures it and what they can do with it. We just say: it's whatever `sublist()` on the matched list returns.

I can't imagine any reasonable person doing nested list patterns inside a rest pattern, which in turn contains a rest pattern. If that's marginally slower than it could be, I won't lose sleep.

I do strongly like that using `sublist()` makes the specification of this and the semantics fairly short and simple. It's essentially just another subpattern that calls `sublist()` instead of `[]`.